qiime2/sdk/usage.py

```diff
@@ -0,0 +1,368 @@
+                             (missing, ))
+
+        extra = provided - exp_outputs
+        if len(extra) > 0:
+            raise ValueError('SDK implementation has specified extra '
+                             'output(s): %r' % (extra, ))
+
+    def build_opts(self, action_signature, scope):
+        opts = {}
+
+        for output in action_signature.outputs.keys():
+            opts[self.get(output)] = output
+
+        return opts
+
+
+class ScopeRecord:
+    STATE_PENDING = 'pending'
+    STATE_FINISHED = 'finished'
+
+    def __init__(self, id: str, factory: callable = None,
+                 value: object = None,
+                 assert_has_line_matching: callable = None):
+        if factory is not None and not callable(factory):
+            # TODO
+            raise Exception
+
+        if assert_has_line_matching is not None and \
+                not callable(assert_has_line_matching):
+            # TODO
+            raise Exception
+
+        self.id = id
+        if value is None:
+            self.state = self.STATE_PENDING
+        else:
+            self.state = self.STATE_FINISHED
+        self._factory = factory
+        self._result = value
+        self._assert_has_line_matching_ = assert_has_line_matching
+
+    def __repr__(self):
+        if self.state == self.STATE_PENDING:
+            return 'ScopeRecord<id=%s, state=%s>' % (self.id, self.state)
+        else:
+            return 'ScopeRecord<id=%s, result=%r>' % (self.id, self.result)
+
+    @property
+    def result(self):
+        if self.is_pending:
+            raise Exception
+        return self._result
+
+    @property
+    def is_pending(self):
+        return self.state == self.STATE_PENDING
+
+    def resolve(self):
+        if self.is_pending:
+            self._result = self._factory()
+            self.state = self.STATE_FINISHED
+
+    def assert_has_line_matching(self, label, path, expression):
+        return self._assert_has_line_matching_(self.id, label, path,
+                                               expression)
+
+
+class Scope:
+    def __init__(self):
+        self._records = dict()
+
+    def __repr__(self):
+        return '%r' % (self._records, )
+
+    @property
+    def records(self):
+        return types.MappingProxyType(self._records)
+
+    def push_record(self, ref, factory=None, value=None,
+                    assert_has_line_matching=None):
+        record = ScopeRecord(id=ref, factory=factory, value=value,
+                             assert_has_line_matching=assert_has_line_matching)
+        self._records[ref] = record
+        return record
+
+    def get_record(self, id):
+        try:
+            return self.records[id]
+        except KeyError:
+            raise KeyError('No record with id: "%s" in scope.' % (id,))
+
+
+class Usage(metaclass=abc.ABCMeta):
+    def __init__(self):
+        self._scope = Scope()
+
+    def init_data(self, ref, factory):
+        override = self._factory_override(ref, factory)
```

Comment:
I think we always want `factory` for input, but it would be nice if `_factory_override` could return a single value, then we can ditch the lambdas that people may forget.

Reply:
This might also help Execution usages in that `_factory_override` could just run the factory, causing any failures to be tied to right the line of `init_data` in the stack trace, instead of being deferred.

Reply:
Are you proposing that we construct the factory here? Or just push a "resolved" record right away?

Reply:
The latter, so if `override` isn't callable, we just take it as resolved.